src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs

```diff
@@ -13,18 +13,24 @@ namespace Microsoft.CodeAnalysis
     public static class IncrementalValueProviderExtensions
     {
         // 1 => 1 transform 
-        public static IncrementalValueProvider<TResult> Select<TSource, TResult>(this IncrementalValueProvider<TSource> source, Func<TSource, CancellationToken, TResult> selector) => new IncrementalValueProvider<TResult>(new TransformNode<TSource, TResult>(source.Node, selector.WrapUserFunction()));
+        public static IncrementalValueProvider<TResult> Select<TSource, TResult>(this IncrementalValueProvider<TSource> source, Func<TSource, CancellationToken, TResult> selector) => new IncrementalValueProvider<TResult>(new TransformNode<TSource, TResult>(source.Node, selector.WrapUserFunctionWithCompilation()));
+        public static IncrementalValueProvider<TResult> Select<TSource, TResult>(this IncrementalValueProvider<TSource> source, Func<TSource, Compilation, CancellationToken, TResult> selector) => new IncrementalValueProvider<TResult>(new TransformNode<TSource, TResult>(source.Node, selector.WrapUserFunction()));
```

Comment:
Here's what this is trying to solve.  Right now if a provider needs access to a compilation, it needs to feed in `context.Compilation` in combined with something else.  However, this is problematic as of course any change to the compilation will then run all later phases since the compilation always changes.  

This means the IG cannot be structured as:

1. use a SyntaxProvider to filter down to basic stuff using nodes+compilations
2. run expensive work using the real compilation later if the something has actually changed.

Effectively, an SG today can only use the compilation in stage one.  But then must compute everything it needs in that stage for later.  Or it needs to create enough of a model of the world that it can do expensive stuff later *absent* any access to compilations/etc.  This means later phases can't use core simple concepts like symbols/etc. as they cannot be passed through providers.


To address this, this PR adjusts things so that the compilation can be accessed at individual phases without rerunning if the compilation itself changes.  This allows basic data to be stored (ints, strings, SymbolKeys, etc.).  If nothing changes with that, nno need to run later phases.  However, if something does change, it can be passed along and merged with the latest compilation to produce the next phase of outputs.

Reply:
The proposed change is _by definition_ erroneous (output depends on undeclared inputs); a code generator using this API would need to be corrected to not rely on this API before it could be used.

Reply:
> output depends on undeclared inputs

That's not true.  The IG is stating that it's results will always be deterministically produced.  But that it can leverage compiler constructs in the generation stage without having to use them at first.

For example, an initial stage can use GetTypeByMetadataName to get and check a type.  Then, later during generation it can go retrieve that type as appropriate for conveniences that our model provides.

Reply:
> The IG is stating that it's results will always be deterministically produced. But that it can leverage compiler constructs in the generation stage without having to use them at first.

This suffers from the same problem described here:
https://github.com/dotnet/roslyn/discussions/59105#discussioncomment-2060075

The incremental pipeline cannot inject data separately from incremental value tracking without breaking its own correctness requirements.

Reply:
i'm not interested in correctness at the cost of making reasonable scenarios nonviable.

Reply:
Due to the impact on correctness, this feature (if implemented) would behave one way inside the IDE, but force the source generator to be ineligible for incrementality when executed in VBCSCompiler. This is viable in theory for the IDE, but would negate all benefits that IIncrementalGenerator might otherwise provide to build, debugging, and/or hot-reload scenarios.

Reply:
Only if hot-reload itself needs perfect correctness.  As with most things, i believe it would likely be fine in practice, esp. as the non-correctness isn't guarantted, only *possible*.  As that's teh state of hte world today anyways (SGs can have bugs after all) that's fine to me.

It does require more care.  But the perf benefits can be substantial.  This can be debated as to whether the perf benefits are worth it or not.

--

e.g. if you ask me:

hey... would you take a 1000x improvement in perf, with a 0.1% chance of an issue, but it was @YairHalberstadt who wrote it, so it's likely very solid.  I would absolutely take that option.